csharp/Adapter/Microsoft.Spark.CSharp/Core/PriorityQueue.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Spark.CSharp.Core
+{
+    /// <summary>
+    /// A bounded priority queue implemented with max binary heap.
+    /// 
+    /// Construction steps:
+    ///  1. Build a Max Heap of the first k elements.
+    ///  2. For each element after the kth element, compare it with the root of the max heap,
+    ///    a. If the element is less than the root, replace root with this element, heapify.
+    ///    b. Else ignore it.
+    /// </summary>
+    [Serializable]
+    internal class PriorityQueue<T> : IEnumerable<T> where T : IComparable<T>
```

Comment:
@tawan0109 - indirect test coverage is better than not having any coverage at all but not a substitute for unit tests providing direct coverage. Consider this scenario - someone modifying just the implementation of PriorityQueue class need a way to verify their changes and ensure the existing functionality is not broken. Without specific unit tests for this class, there is no straightforward way.  They have to find out the usage of the PriorityQueue, locate the corresponding unit test and rely on them. Even that cannot provide the confidence like direct unit tests. Just wanted to share my 2 cents on this.